plugin/templates/resource_name.py

```diff
@@ -110,14 +112,113 @@ def template_name(self):
+        ]
+        if self.format_fields:
+            self.format_fields[0]['not_first'] = False
+            self.format_fields[-1]['not_last'] = False
+
+        self.patterns = []
+        for p in pattern_strings:
+            pattern_name_lower_underscore = \
+                get_pattern_name(p)
+            pattern_name_lower_camel = \
+                casing_utils.lower_underscore_to_lower_camel(
+                    pattern_name_lower_underscore)
+            pattern_name_upper_camel = \
+                casing_utils.lower_underscore_to_upper_camel(
+                    pattern_name_lower_underscore)
+            pattern_name_upper_underscore = \
+                casing_utils.lower_underscore_to_upper_underscore(
+                    pattern_name_lower_underscore)
+
+            pattern_format_fields = \
+                get_format_fields_for_pattern(p,
+                                              pattern_to_id_segments,
+                                              segment_to_segment_symbols)
+
+            self.patterns.append(
+                ResourceNamePattern(
+                    p,
+                    pattern_name_lower_camel,
+                    pattern_name_upper_camel,
+                    pattern_name_upper_underscore,
+                    pattern_format_fields))
+
+        if len(self.patterns) > 0:
+            self.first_pattern = self.patterns[0]
+            self.patterns[0].set_first()
+            self.patterns[0].set_short_builder_name()
+            self.patterns[-1].set_last()
 
     def template_name(self):
-        return "parent_resource_name.mustache"
+        return "multi_pattern_resource_name.mustache" if self.patterns \
+            else "deprecated_parent_resource_name.mustache"
+
+
+class ResourceNamePattern:
+
+    def __init__(self, pattern_string,
+                 pattern_name_lower_camel,
+                 pattern_name_upper_camel,
+                 pattern_name_upper_underscore,
+                 format_fields):
+        self.is_fixed = len(format_fields) == 0
+        self.is_formattable = not self.is_fixed
+        self.pattern_string = pattern_string
+        self.upper_camel = pattern_name_upper_camel
+        self.lower_camel = pattern_name_lower_camel
+        self.upper_underscore = pattern_name_upper_underscore
+        self.format_fields = format_fields
+        self.builder_name = self.upper_camel + 'Builder'
+        if format_fields:
+            self.format_fields[0]['not_first'] = False
+            self.format_fields[-1]['not_last'] = False
+        for format_field in self.format_fields:
+            format_field['pattern_builder_name'] = self.builder_name
+        self.not_first = True
+        self.is_first = False
+        self.not_last = True
+
+    def set_first(self):
+        self.not_first = False
+        self.is_first = True
+
+    def set_last(self):
+        self.not_last = False
+
+    def set_short_builder_name(self):
```

Comment:
What is `short_builder_name`? Also this method seems like modifying the value of `pattern_builder_name` key in format_field dictionary. Is `short_builder_name` and `pattern_builder_name` same thing?

Reply:
So the first `Builder` class should be called `Builder` instead of `XXXBuilder` (e.g., `ProjectLocationBookBuilder`) for backward-compat when a resource evolves from single-pattern to multi-pattern, because a single resource name's builder class is always called `Builder`. This method sets this `ResourceNamePattern` to use the short builder name (`Builder`).